File: Lab Final/src/main/java/org/example/service/IssueService.java
package org.example.service;

import org.example.command.*;
import org.example.model.Comment;
import org.example.model.Issue;
import org.example.model.Priority;
import org.example.model.Status;
import org.example.repository.IssueRepository;

import java.util.Collection;
import java.util.Optional;

public class IssueService {
    private final IssueRepository repository;
    private final CommandManager commandManager;

    public IssueService(IssueRepository repository, CommandManager commandManager) {
        this.repository = repository;
        this.commandManager = commandManager;
    }

    public Issue createIssue(String title, String description, Priority priority) {
        Issue issue = new Issue(title, description, priority);
        CreateIssueCommand command = new CreateIssueCommand(issue, repository);
        commandManager.executeCommand(command);
        return issue;
    }

    public void updateIssue(Issue issue) {
        UpdateIssueCommand command = new UpdateIssueCommand(issue, repository);
        commandManager.executeCommand(command);
    }

    public void changeIssueStatus(String issueId, Status newStatus) {
        ChangeIssueStatusCommand command = new ChangeIssueStatusCommand(issueId, newStatus, repository);
        commandManager.executeCommand(command);
    }

    public void addComment(String issueId, String content, String author) {
        Comment comment = new Comment(content, author);
        AddCommentCommand command = new AddCommentCommand(issueId, comment, repository);
        commandManager.executeCommand(command);
    }

    public Optional<Issue> getIssue(String id) {
        return repository.findById(id);
    }

    public Collection<Issue> getAllIssues() {
        return repository.findAll();
    }

    public void undo() {
        commandManager.undo();
    }

    public void redo() {
        commandManager.redo();
    }

    public boolean canUndo() {
        return commandManager.canUndo();
    }

    public boolean canRedo() {
        return commandManager.canRedo();
    }
}

File: TaskManager/TaskManager/Service/IssueService.cs
using TaskManager.Commands;
using TaskManager.Domain;
using TaskManager.Repository;
using TaskManager.Subjects;

namespace TaskManager.Service;

public class IssueService
{
    private readonly IssueRepository _repository;
    private readonly CommandManager _commandManager;

    public IssueService(IssueRepository repository, CommandManager commandManager)
    {
        _repository = repository;
        _commandManager = commandManager;
    }

    public int CreateIssue(string title, string description, IssuePriority priority, string assignedTo = null)
    {
        var issue = new Issue
        {
            Title = title,
            Description = description,
            Priority = priority,
            AssignedTo = assignedTo
        };

        var command = new CreateIssueCommand(_repository, issue);
        _commandManager.ExecuteCommand(command);
        return issue.Id;
    }

    public void UpdateIssue(Issue issue)
    {
        var command = new UpdateIssueCommand(_repository, issue);
        _commandManager.ExecuteCommand(command);
    }

    public void AddComment(int issueId, string content, string author)
    {
        var comment = new Comment
        {
            Content = content,
            Author = author
        };
        var command = new AddCommentCommand(_repository, issueId, comment);
        _commandManager.ExecuteCommand(command);
    }

    public void ChangeStatus(int issueId, IssueStatus status)
    {
        var command = new ChangeIssueStatusCommand(_repository, issueId, status);
        _commandManager.ExecuteCommand(command);
    }

    public Issue GetIssue(int id)
    {
        return _repository.GetById(id);
    }

    public IEnumerable<Issue> GetAllIssues()
    {
        return _repository.GetAll();
    }

    public void UndoLastCommand()
    {
        _commandManager.Undo();
    }

    public void RedoCommand()
    {
        _commandManager.Redo();
    }

    public IEnumerable<ICommand> GetCommandHistory()
    {
        return _commandManager.GetCommandHistory();
    }
}